Static proot, wafHook cross compilation

symphorien opened the pull request:
###### Motivation for this change

Have `pkgsStatic.proot.override {enablePython = false;}` compile.

###### Things done

* Let `wafHook` use the vendored waf even if it is not located at `./waf`
* Make `wafHook` ignore `--disable-shared` added by pkgsStatic
* Make `wafHook` add the necessary configure flags for cross compilation transparently
* adapt tdb and talloc to these modifications
* fix cross compilation of proot

###### Things tested

Compilation of `tdb`, `talloc` and `proot` (with python disabled) in `pkgs`, `pkgsStatic` and `pkgsCross.raspberryPi`. `pkgsStatic.tdb` does not work because the waf script can only build a shared library.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).


cc @matthewbauer as the author of wafHook.

symphorien committed:
wafHook: let derivations use an existing waf in a non-standard location

this is necessary for talloc

symphorien committed:
wafHook: ignore --disable-static and such options

These options are forcefully added by pkgsStatic but are not understood
by waf.

symphorien committed:
wafHook: ignore configurePlatforms

waf does support --build and --host, but the only effect of these
options is an error message telling to use --cross-compile instead.
So we ignore these flags.

symphorien committed:
wafHook: cross compilation support

for this to work, wafHook must be in nativeBuildInputs.

symphorien committed:
talloc, tdb: fix cross-compilation after wafHook improvements

symphorien committed:
proot: fix cross compilation

matthewbauer commented on pkgs/development/tools/build-managers/waf/setup-hook.sh:
```diff
@@ -1,20 +1,40 @@
+        if [[
+        # waf does not support these flags, but they are "blindly" added by the
+        # pkgsStatic overlay, for example.
+              $flag != "--enable-static"
```
Perhaps we should have some separate variable like "wafFlags" for this? waf supports a subset of the configure flags but not enough apparently.

symphorien replied:
This would be a more invasive change. Would you prefer I change the PR to do this?

matthewbauer replied:
It's okay as is for now. This is more a note on this.

matthewbauer commented on pkgs/top-level/all-packages.nix:
```diff
@@ -6366,7 +6366,11 @@ in
+    substitutions = {
+      inherit waf;
+      crossFlags = lib.optionalString (stdenv.hostPlatform != stdenv.targetPlatform)
+        ''--cross-compile "--cross-execute=${stdenv.targetPlatform.emulator pkgs}"'';
```
Note that this is pretty expensive to build. I'm not sure if waf is used enough for it to a problem though. I think it's okay since all waf scripts support it (not all need it though).

symphorien replied:
If I understand correctly, this will only be built when one cross-builds a derivation using wafHook, so this is pretty optimal.
```
~/src/nixpkgs static-proot* 
$   nix-store -qR $(nix-instantiate . -A pkgsCross.raspberryPi.talloc) | grep qemu
warning: you did not specify '--add-root'; the result might be removed by the garbage collector
/nix/store/7s9kxar0nh6d6nzri342ynpfpxniwq1b-fix-qemu-ga.patch
/nix/store/pgkjdxzpglzfj5hpjim1p1m8bipx9hmj-qemu-3.1.0.tar.bz2.drv
/nix/store/3pjswv7hpf0z4avx11dp16rb9wfwpppg-qemu-3.1.0.drv
                                                                                                                                                                                                       
~/src/nixpkgs static-proot* 
$   nix-store -qR $(nix-instantiate . -A talloc) | grep qemu 
warning: you did not specify '--add-root'; the result might be removed by the garbage collector
```

matthewbauer replied:
Yeah that's right. There might be a benefit to only adding it when you know a script will need to cross execute. But there's not a clean way to do that in Nixpkgs right now.

matthewbauer commented on pkgs/development/tools/build-managers/waf/setup-hook.sh:
```diff
@@ -1,20 +1,40 @@
-    local flagsArray=(
-        $configureFlags ${configureFlagsArray[@]}
+    local flagsArray=(@crossFlags@);
+    for flag in $configureFlags "${configureFlagsArray[@]}";
```
Some more flags we might want to add are ``` --bundled-libraries=NONE``` and ``` --disable-rpath-install``` according to https://wiki.samba.org/index.php/Waf#controlling_rpath

symphorien replied:
Makes sense, but this should probably go in another PR.

Ekleog commented:
@matthewbauer Any reason not to merge this? I assume people who would have objections would have raised in the 11 days since your r+ :)

veprbl reviewed (commented):
I think this should really change to use `wafFlags` because ignoring flags is a hack and waf is not a build system with a defined set of flags, developers are free to define their own `--enable-static` option. Keeping such hack would also mean that we need to maintain it with respect to all changes to `pkgsStatic`.

symphorien commented:
`wafFlags` is already taken for build flags, shall I go with`wafConfigureFlags`?

matthewbauer committed:
waf: use wafConfigureFlags for waf flags

This avoids the potential conflict between autoconf flags and the waf
flags. There is some overlap between the two but waf errors when it
doesn’t recognize the flag.

matthewbauer commented:
@symphorien I pushed a commit to do this. Feel free to drop it if it's not helpful!

symphorien commented:
Nice, thanks. It might warrant some release notes. I'll try to have a closer look on the week-end.

veprbl commented:
@GrahamcOfBorg build proot

veprbl commented:
@GrahamcOfBorg build talloc
